tensorflow/compiler/mlir/hlo/lib/Analysis/userange_analysis.cc

```diff
@@ -291,6 +285,47 @@ void UseInterval::intervalSubtract(UseInterval::Vector &a,
   }
 }
 
+/// Performs an interval intersection => A = A ^ B.
+void UseInterval::intervalIntersect(UseInterval::Vector &a,
+                                    const UseInterval::Vector &b) {
+  auto iterB = b.begin();
+  auto endB = b.end();
+  for (auto iterA = a.begin(); iterA != a.end() && iterB != endB;) {
+    // iterA is strictly before iterB => erase iterA.
+    if (*iterA < *iterB) {
+      iterA = a.erase(iterA);
+      // iterB is strictly before iterA => increment iterB.
+    } else if (*iterA > *iterB) {
+      ++iterB;
+      // iterA overlaps with the start of iterB. Set the start of iterA to the
+      // start of iterB and increment iterA.
+    } else if (iterA->start < iterB->start && iterA->end <= iterB->end) {
```

Comment:
You can combine these two cases. In both you reduce the current interval to the overlap. If the split-off is non-empty, you add it to the iteration.